### PR TITLE
chore: migrate withViewportOffsetTop.js class to function component

### PR DESCRIPTION
**Pull request by teneeto**
### Details

This change is made to migrate the [withViewportOffsetTop.js](https://github.com/Expensify/App/blob/main/src/components/withViewportOffsetTop.js) from class to a function component

### Fixed Issues

$: [16224](https://github.com/Expensify/App/issues/16224)
PROPOSAL:  [16224](https://github.com/Expensify/App/issues/16224)

### Tests

1. Open any chat
2. Focus the chat composer
3. Make sure you can still see the header

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Open any chat
2. Focus the chat composer
3. Make sure you can still see the header

### QA Steps

1. Open any chat
2. Focus the chat composer
3. Make sure you can still see the header

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/22779249/b19c3750-9b48-4ef5-b73a-09f96fbbf4b1

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/22779249/09957aa3-5b36-4c05-919b-05137402a7b2

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/22779249/e0d11ad5-a09f-4ecb-8cd4-27577321a7b9

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/22779249/e96a6349-ffd9-4701-90f1-54c03b58bee7

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/22779249/a00353ab-390f-4bb2-befc-8c194f1d72fa

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/22779249/e45c4000-5e6d-4924-9dd4-c47975ff05a8

</details>

### CODE REVIEW

File: src/components/withViewportOffsetTop.js [hunk 1]
```diff
@@ -13,43 +13,33 @@ const viewportOffsetTopPropTypes = {
...
          * @param {SyntheticEvent} e
          */
-        updateDimensions(e) {
-            const viewportOffsetTop = lodashGet(e, 'target.offsetTop', 0);
-            this.setState({viewportOffsetTop});
-        }
+        const updateDimensions = useCallback((e) => {
+            const viewPortOffsetTop = lodashGet(e, 'target.offsetTop', 0);
+            setViewportOffsetTop(viewPortOffsetTop);
+        }, []);
```
Comment (thiagobrez) [inline src/components/withViewportOffsetTop.js:25]: Seems like this function is only used for the listener below. We can move it inside the `useEffect`, then we have no need to wrap it in `useCallback`, and the effect dependencies will also be cleared

**Review by amyevans — APPROVED** [review]

**Comment by amyevans** [comment]
This is still in need of a cross-agency Reviewer checklist, will wait to merge until that's set!

**Comment by situchan** [comment]
C+ can review and checklist for function component migration PRs.
Cross-agency Reviewer checklist is only for TS migration PRs.

**Comment by amyevans** [comment]
Ah right, thanks for the reminder @situchan!

@thesahindia you're the assigned C+ on the issue, are you able to review?

**Comment by teneeto** [comment]
> Hey @teneeto, can you please add some kind of test steps?

@thesahindia I have updated the description with the reference link of the initial PR description that authored/created the [`withViewportOffsetTop`](https://github.com/Expensify/App/pull/14725/commits/2fc343c98bd96c91a2ee804a4cc6d34448aa518f)  HOC. There are detailed steps to test and qa [here](https://github.com/Expensify/App/pull/14725#issue-1565689464). Thanks.

**Comment by teneeto** [comment]
> > 1. Go to Settings -> Payment
> > 2. Tap on the bank account that doesn't have the "default" label
> > 3. Tap on "Make default payment method" button
> > 4. Ensure the keyboard does not overlap the "Make default payment method" button.
> 
> It looks like now we don't ask for the password while trying to set the payment method as default so there is no input field and the keyboard can't be opened, so we can't follow these steps.
> 
> We are using `viewportOffsetTop` at [ReportScreen](https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/pages/home/ReportScreen.js#L162) and [EmojiPicker](https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/components/EmojiPicker/EmojiPicker.js#L164), I tried looking at PRs that added them but couldn't find info about why we need them.
> 
> The code changes looks good to me but I wanted to check if things are working fine. Not sure what to do next.
> 
> cc: @amyevans

let me take a look at another possible qa step.

**Comment by teneeto** [comment]
@thesahindia 
1. Goto Settings -> Profile
2. Tap on Personal Details
3. Tap on Legal name
4. TAP on the `Legal first name` input

Note: Notice how the keyboard pushes the save button rather than overlap.


https://github.com/Expensify/App/assets/22779249/aaeb7a6a-3c2a-4e29-9c4d-55227df83927

**Comment by teneeto** [comment]
> @thesahindia
> 
> 1. Goto Settings -> Profile
> 2. Tap on Personal Details
> 3. Tap on Legal name
> 4. TAP on the `Legal first name` input
> 
> Note: Notice how the keyboard pushes the save button rather than overlap.
> 
>  Screen.Recording.2023-10-02.at.05.41.54.mov

Let me know if this works for you and I'll update the test/qa steps.

**Comment by teneeto** [comment]
Just a friendly reminder! @thesahindia can we kindly have a look at this again with the updated qa steps.

**Comment by teneeto** [comment]
> @teneeto, sorry but I am little confused. We are using `viewportOffsetTop` at [ReportScreen](https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/pages/home/ReportScreen.js#L162) and [EmojiPicker](https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/components/EmojiPicker/EmojiPicker.js#L164). So I am not sure how it affects legal name page.

Ayye! Let me check again....

**Comment by teneeto** [comment]
> > 1. Go to Settings -> Payment
> > 2. Tap on the bank account that doesn't have the "default" label
> > 3. Tap on "Make default payment method" button
> > 4. Ensure the keyboard does not overlap the "Make default payment method" button.
> 
> It looks like now we don't ask for the password while trying to set the payment method as default so there is no input field and the keyboard can't be opened, so we can't follow these steps.
> 
> We are using `viewportOffsetTop` at [ReportScreen](https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/pages/home/ReportScreen.js#L162) and [EmojiPicker](https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/components/EmojiPicker/EmojiPicker.js#L164), I tried looking at PRs that added them but couldn't find info about why we need them.
> 
> The code changes looks good to me but I wanted to check if things are working fine. Not sure what to do next.
> 
> cc: @amyevans

@thesahindia this seem to be the only visible way to test this. But again, since the ViewPortOffsetTop is used within EmojiPicker, the component is called each time you're in a chat screen. can tell visibly exactly what it solves, but i still think it's part of a solution as referenced [here](https://github.com/Expensify/App/pull/14725#issue-1565689464).

@amyevans @thesahindia how do we proceed with this?

**Comment by amyevans** [comment]
If you read [the proposal](https://github.com/Expensify/App/issues/14003#issuecomment-1410056441) linked from [the PR that implemented this component](https://github.com/Expensify/App/pull/14725), it says:

> The same issue occurs on the reportScreen, the way that was fixed there is by using the visualViewport listener and adding [padding to the top of ScreenWrapper](https://github.com/Expensify/App/blob/8831e76e40821a7174f7e31987d5d2027859882b/src/pages/home/ReportScreen.js#L218).

If you then follow the git blame history, you can get to this PR: https://github.com/Expensify/App/pull/10208, where the test steps are:
1. Open any chat
2. Focus the chat composer
3. Make sure you can still see the header

mWeb Safari seems to be the crucial platform to check for a regression in that case.

With `const screenWrapperStyle = [styles.appContent, styles.flex1, {marginTop: viewportOffsetTop}];`:
Screenshot 2023-10-10 at 3 27 31 PM

With `const screenWrapperStyle = [styles.appContent, styles.flex1];`:
Screenshot 2023-10-10 at 3 27 53 PM

**Review by thesahindia — APPROVED** [review]
LGTM!

**Comment by teneeto** [comment]
Steps Updated!

**Pull request merged**